mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td

```diff
@@ -1370,10 +1370,17 @@ def ModuleFlagCGProfileEntryAttr
           ]>]
     ```
   }];
-  let parameters = (ins "FlatSymbolRefAttr":$from,
-                        "FlatSymbolRefAttr":$to,
-                        "uint64_t":$count);
-  let assemblyFormat = "`<` struct(params) `>`";
+  let parameters = (
+    ins OptionalParameter<"std::optional<FlatSymbolRefAttr>">:$from,
+        OptionalParameter<"std::optional<FlatSymbolRefAttr>">:$to,
```

Comment:
Wouldn't using OptionalParameter<"FlatSymbolRefAttr"> on its own be enough? For attributes, the nullptr can be used as optional value and AFAIK the printer and parser by default will just omit the nullptr values. That way the custom parser and printer should not be necessary except you want to print the null value explicitly.

Reply:
The problem is that we either get:
- `to = @<<INVALID EMPTY SYMBOL>>` for passing an empty string to `FlatSymbolRefAttr`, or
- `to =, count = 2` for passing `nullptr`. I could live with this, but `to = null, count = 2` feels nicer to me. 

When I use `OptionalParameter<"FlatSymbolRefAttr">`, the printer works successfully but parser fails on `to = null` with `failed to parse ModuleFlagCGProfileEntryAttr parameter 'to' which is to be a FlatSymbolRefAttr`. Seems like a current limitation of emitted tablegen code? Perhaps this explains why some other places also do the double optional trick. If you feel stronger about other options let me know and will fix them in a follow up PR.

Reply:
> When I use OptionalParameter<"FlatSymbolRefAttr">, the printer works successfully but parser fails on to = 

Yeah I would expect that instead of `to = null` you would have to omit the to to get a nullptr.

e.g instead of:
```
 #llvm.cgprofile_entry<from = null, to = @to, count = 222>,
 #llvm.cgprofile_entry<from = @from, to = null, count = 222>,
``
I would expect that the attributes without std:optional would print/parse like this:
```
 #llvm.cgprofile_entry<to = @to, count = 222>,
 #llvm.cgprofile_entry<from = @from, count = 222>,
```
meaning the null symbol references are omitted. 

If you can get it working, it is strongly preferable over implementing our own printer and parser even if the IR looks a bit different than in LLVM IR.

Reply:
Oh, I see your point, that should work, let me try it out

Reply:
finally solved!